LogicLayer/Friends/FriendsService.cs

```diff
@@ -54,16 +59,17 @@ public async Task SendFriendRequestAsync(CreateFriendRequestDto friendRequest)
                 throw new InvitingYourselfException("You cannot invite yourself.");
             }
 
-            var requestedById = await _repository.GetUserIdByUserName(friendRequest.UserNameFrom);
-            var requestedToId = await _repository.GetUserIdByUserName(friendRequest.UserNameTo);
+            var userRequestedBy = _repository.ApplicationUsers.Single(u => u.UserName == friendRequest.UserNameFrom);
+            var userRequestedTo = _repository.ApplicationUsers.Single(u => u.UserName == friendRequest.UserNameTo);
```

Comment:
If you're searching by user name you should be searching by NormalizedUserName.

Reply:
Done?

Reply:
You're comparing an unormalized username to the normalized one. Look at line 30, you need `ToUpper`. Also, this will result in code duplication, so extract finding user by username to a method somewhere (extension method on the `ApplicationUser` that returns `true` if the passed username is the same after normalization?)

Reply:
Look if it gets 51.

Reply:
No, this is terribly inefficient now. If you look at the app output it throws warnings stating "`UserNameEquals` could not be translated and will be evaluated locally." After that everything fails to be translated to SQL, because everything depends on that user query, so EF proceeds to load the entire `User` table and the entire `Friendship` table, and then the `User` tables again, and it all happens several times. Standard methods cannot be translated into SQL by LINQ, so an extension method will not do. You need something that can be compiled into an expression tree, so of type `Expression<Func<ApplicationUser, bool>>` in this case. 

```csharp
static Expression<Func<ApplicationUser, bool>> UserNameEquals(string userName) =>
    u => u.NormalizedUserName == userName;

_repository.ApplicationUser.Where(UserNameEquals(userName));
```

Reply:
I fixed this by myself and forced LINQ to throw exceptions when stuff like this happens. I'd like you @Johny7013 to look at the changes in my last commit and review them. If you have issues, comment here, if it's a-ok then resolve this conversation.